### ISSUE TITLE
Supporting EQL in Elasticsearch

### ISSUE THREAD

**Issue by colings86** [elastic/elasticsearch#49581]
This is a meta issue to track progress of adding EQL support to Elasticsearch. EQL will be supported via a new Elastic licensed plugin which will provide execution of EQL rules.

The language reference for EQL can be found [here](https://eql.readthedocs.io/en/latest/query-guide/index.html)

# Scope of first iteration

### In Scope
* Sequences
* Joins (Note this is specifically EQL joins which are most like a merge join)
* Pipes
* Most functions (see https://github.com/elastic/elasticsearch/issues/51556)

### Out of Scope
* Lineage, i.e.:
    * Descendant of
    * Child of
    * Event of
* Array functions
    * arrayContains
    * arraySearch
    * arrayCount
* `fork` in sequences

# Things to consider during the design and implementation
* EQL is a language that’s already used in the wild. Any changes we make to the language or the way EQL rules are written will have implications on the existing users and migration of users to future versions so these should be very carefully considered
* A separate implementation of EQL will still exist that queries the endpoint devices directly. This means that we should avoid differences between writing EQL for Elasticsearch and EQL to run on the endpoints as any differences would have the potential to confuse users
* We see EQL being useful in many use cases. The Elasticsearch implementation need to be able to be used on generic data not just Endpoint data. EQL should default to expecting data in ECS format, but should have the ability to be overridden so it can be used on other schemas as well

# High level tasks
Each task here has its own issue and some bigger tasks might have their own meta issue:
* [x] Create EQL plugin https://github.com/elastic/elasticsearch/issues/49583
* [x] Extract reusable components from SQL into common project https://github.com/elastic/elasticsearch/issues/49773
* [x] Synchronous EQL REST API for ad-hoc querying https://github.com/elastic/elasticsearch/issues/49634
* [x] Build query parser and plan on reusable SQL components https://github.com/elastic/elasticsearch/issues/49997
* [x] EQL transpiler for stateless EQL expressions to ES Search DSL https://github.com/elastic/elasticsearch/issues/49589
* [x] **MILESTONE 1: Ability to run stateless EQL queries from a synchronous request/response API**
    * This means running queries of the form `event_type where CRITERIA` where `CRITERIA` can use `field_name=value`, together with and/or/not
   * The REST API will be a synchronous request/response so the client will have to wait on the connection for the response
* [x] Response format enhancements https://github.com/elastic/elasticsearch/issues/52845
* [x] Support existing EQL functions https://github.com/elastic/elasticsearch/issues/51556
* [x] Map EQL sequence parts to ES requests https://github.com/elastic/elasticsearch/issues/49590
* [ ] Sequence and join execution https://github.com/elastic/elasticsearch/issues/49594
* [ ] Implement pipes logic in the plugin (not using aggregations) https://github.com/elastic/elasticsearch/issues/49627
* [x] Add support for EQL functions #51556
* [x] High Level REST client support for EQL API https://github.com/elastic/elasticsearch/issues/51961
* [x] Cancelling a task's grandchildren when the task is cancelled #50990
* [x] Convert EQL REST API to be async (similar to async search) https://github.com/elastic/elasticsearch/issues/49638
* [x] Create documentation for using EQL #51057
* [ ] **MILESTONE 2: Supports all in scope EQL language features for first release in an asynchronous API**
    * This means the plugin will functionally support all the elements int he "Language Features" section above"
   * The API will be asynchronous so a client can send a request and have a task ID returned, then check back for progress and final results later
   * Although functionally complete there are still items below we would want to implement before the first release is GA ready
* [x] Develop Point-in-time reader @jimczi https://github.com/elastic/elasticsearch/issues/46523
* [x] Modify EQL execution to use point in time reader https://github.com/elastic/elasticsearch/issues/49628
* [ ] Push down pipes to aggregations where possible https://github.com/elastic/elasticsearch/issues/49629
* [x] Telemetry for EQL usage https://github.com/elastic/elasticsearch/issues/49630
* [x] Improvements to search_after performance https://github.com/elastic/elasticsearch/issues/37043
* [x] Improve wildcard search performance https://github.com/elastic/elasticsearch/issues/48852
* [x] Escape non-alphanumeric fields for EQL #51443
* [ ] **MILESTONE 3: Ready for first release**
* [x] Remove feature flags and prepare for release https://github.com/elastic/elasticsearch/issues/51613

**Comment by alexbrasetvik** [comment]
It seems that some of the infrastructure required for this (notably the multiple search request executor and limited join-handling) could also eventually help support SQL `WITH`-queries (common table expressions) in some distant future :)

**Comment by costin** [comment]
Replacing this long standing issue with the remaining tasks for 7.11 at https://github.com/elastic/elasticsearch/issues/64273
